fix race condition whereby physics may incorrectly overwrite script/network property change

AndrewMeadows opened the pull request.

AndrewMeadows commented:
TODO: add comments and test plan

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6341/app_12185_PR.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6341/HighFidelity-Beta-PR12185-4df5290237ab476b922eb726014ec15f819b916f.dmg

AndrewMeadows commented:
Jenkins, rebuild this please.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6345/app_12185_PR.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6345/HighFidelity-Beta-PR12185-4df5290237ab476b922eb726014ec15f819b916f.dmg

marko8904 commented:
build this please

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6349/app_12185_PR.apk

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6349/HighFidelity-Beta-PR12185-4df5290237ab476b922eb726014ec15f819b916f.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6349/HighFidelity-Beta-PR12185-4df5290237ab476b922eb726014ec15f819b916f.dmg

sethalves commented on libraries/entities/src/EntityItem.cpp:
```diff
@@ -1610,6 +1610,37 @@ void EntityItem::setPosition(const glm::vec3& value) {
+        if (!(flags & Simulation::DIRTY_ANGULAR_VELOCITY)) {
+            setWorldAngularVelocity(angularVelocity);
+        }
+        if (flags != (Simulation::DIRTY_TRANSFORM | Simulation::DIRTY_VELOCITIES)) {
```
could there be unrelated flags that confound this if?

sethalves replied:
nevermind, I see that flags is limited to these, above.

sethalves commented on libraries/entities/src/EntityItem.cpp:
```diff
@@ -1610,6 +1610,37 @@ void EntityItem::setPosition(const glm::vec3& value) {
+        setWorldAngularVelocity(angularVelocity);
+        setLastSimulated(usecTimestampNow());
+    } else {
+        // only set properties NOT flagged
```
this code makes me think you don't trust your optimizer.

AndrewMeadows replied:
Indeed, I am guilty.

AndrewMeadows commented:
Test Plan:

**(1)** Login to a domain where you can change add and edit entities.  It helps if you're using a PC that is CPU-challenged (e.g. i5) or login to a domain that spends a significant time in `PhysicsEngine::stepSimulation()` (plenty of active dynamic objects moving about).

**(2)** Run the following script which will create an object (with 5 min lifetime) that runs an "entity script" via URL that will set the velocity of the object to be upward at 1m/sec every time it is clicked:
```
var orientation = MyAvatar.orientation;
var offsetDistance = 3.0;
var localOffset = { x:0, y: 0, z: -offsetDistance };
var center = Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(orientation, localOffset));
var ballColor = { red: 255, green: 127, blue: 64 };
var scriptUrl = "https://raw.githubusercontent.com/AndrewMeadows/hifi-js/master/example/pushBall.js";
var LIFETIME = 600;

var properties = {
    name: "test-sphere",
    type: "Sphere",
    color: ballColor,
    position: center,
    dimensions: { x: 1, y: 1, z: 1 },
    script: scriptUrl,
    lifetime: LIFETIME
}
var myEntity = Entities.addEntity(properties);

function cleanup() {
    Entities.deleteEntity(myEntity);
}
Script.scriptEnding.connect(cleanup);
```

**(3)** Click on the object severaltimes:

--> on `hifi/master` you should notice that sometimes a click does not actually register and update the object's velocity.  This is because sometimes the physics simulation will overwrite the velocity change made by the script.

--> on this PR each click should ALWAYS update the object's velocity to be <0,1,0> upward.

^^^ This is the simple repro for the bug.  There may be other buggy behaviors that are fixed, such as:

**(a)** Object oscillating between two positions: one set by a physics Action (grab) and another set via some script.
**(b)** Object fails to be updated by second interface's script when first interface owns the simulation.

QAadahlin commented:
Currently Testing

AndrewMeadows commented:
@QAadahlin, sorry there was a bug in the remote pushBall.js script that the rez-script in the test plan registers on the object.  I added a pushBall-003.js to get around the caching problem and updated the test plan script to use that.

QAadahlin commented:
@conklin94122 Testing Completed. PASSED.

ZappoMan commented:
Please wait for my review.

QAadahlin commented:
@conklin94122 @andrewd440 Testing completed. PASSED.

ZappoMan commented on libraries/entities/src/EntityItem.cpp:
```diff
@@ -1610,6 +1610,37 @@ void EntityItem::setPosition(const glm::vec3& value) {
     }
 }
 
+void EntityItem::setWorldTransformAndVelocitiesUnlessDirtyFlags(
```
Why are you adding these as methods on EntityItem, why can't you just call these public methods in the `EntityMotionState::setWorldTransform()` function?

ZappoMan commented on libraries/physics/src/EntityMotionState.cpp:
```diff
@@ -256,25 +256,12 @@ void EntityMotionState::setWorldTransform(const btTransform& worldTrans) {
-    _entity->setAngularVelocity(getBodyAngularVelocity());
-    _entity->setLastSimulated(usecTimestampNow());
+
+    _entity->setWorldTransformAndVelocitiesUnlessDirtyFlags(
```
so from what I can tell, your real change here is that the `EntityMotionState::setWorldTransform()` operation, used to just slam new velocities and transforms onto the entity, but now, they check for the dirty flag state before updating the values. This makes sense. But why are you putting this very germanic word salad of a public function name `setWorldTransformAndVelocitiesUnlessDirtyFlags` onto the EntityItem... when really this job of checking for dirtiness and only overriding them is meaningful only to the motion state object.

The EntityItems area of concern is the storage and management of the properties and state of the entity. The EntityMotionState's area of concern is the simulation and updating of that state specific to the simulation engine.. The task of determining that the state should or should not be updated seems to belong to the area of concern of the EntityMotionState in this case.

AndrewMeadows commented:
The good news is: these changes appear to fix all of the problems with the "crowd beach ball" content.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6382/app_12185_PR.apk

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6383/app_12185_PR.apk

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6383/HighFidelity-Beta-PR12185-47101a072228ccada9c8414e23483730b5102e75.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6383/HighFidelity-Beta-PR12185-47101a072228ccada9c8414e23483730b5102e75.dmg

ZappoMan commented:
@conklin94122 @QAadahlin - needs retest after code reworked.

QAadahlin commented:
Currently testing

QAadahlin commented:
@conklin94122 Completed testing. FAILED.
When I tested it in my sandbox, it functioned as normal. However, when I moved to the Dev-welcome and loaded the script there it at first over-responded by bouncing higher than normal and getting stuck mid-air (see screenshot below for reference). I removed the script and tried it again - only this time it did not respond to touch for the first 10-15 seconds (after which it would respond as normal).  I also noted this behavior in the Eschatology domain. It continued to function as normal in my sandbox. I will also include logs for reference. 

Ball stuck mid-air
![ball bounce](https://user-images.githubusercontent.com/35040758/35109381-7121a41e-fc33-11e7-8f5b-7866ee920f80.JPG)

[hifi-log.txt](https://github.com/highfidelity/hifi/files/1643672/hifi-log.txt)

AndrewMeadows commented:
Hrm so... it works for local domains but not remote ones?  I will test it.

ZappoMan commented:
I am suspicious of the fact that this passed QA before the code review changes, but is failing now. The changes after code review *shouldn't have* impacted behavior, it was just rearranging code. What this suggests to me is that the original test pass that was considered a "pass" was not thorough enough, and that whatever is occurring in this second pass was not covered by the original test. Let's make sure we understand the difference between these two test passes...

conklin94122 commented:
@ZappoMan A mistake was made and the pre and post code review builds were not tested identically. We are going to retest it now, thoroughly, and share results. If the most recent build fails we will attempt to reproduce those fails against the build immediately before the code review, and include those results here.

QAadahlin commented:
Currently testing

AndrewMeadows commented:
I've been testing in Dev-Welcome and can say I'm seeing some strange bugginess,  however what I'm seeing is also happening in the tip of hifi/master (dev-download 7708).  As such, although the bugs are impacting the test-plan, I believe they are independent of the changes in this PR.  Incidentally, things were working fine for me in Eschatology, however I only tested there once and then started trying to figure out what is happening in Dev-Welcome.

What I'm seeing:

(1) When I login and rez the test-sphere sometimes its properties (or something) seems incomplete.  The sphere will not collide with my avatar although when I check its properties they should make it collideable.  It is as if the sphere hasn't been added to the physics engine yet.  Often the sphere will eventually become collidable after time, although sometimes I've seen it suddenly collidable and dynamic... only to shortly stop moving.  When I check the properties again I see that the object is no longer dynamic.

Typically these problems happen right after login, and then stop happening.  After a couple minutes things appear to be working more as expected, but when I logout/login again the problems resume.

(2) I'm noticing a problem where the sphere disappears.  The LOD auto-adjust system is culling it.  This, despite the fact that the render rates are all fine: LOD should recover.  I'll test with PR-12157 which has some improved tools for monitoring the measurements driving LOD auto adjust.

conklin94122 commented:
@ZappoMan @AndrewMeadows

conklin94122 commented:
@ZappoMan @AndrewMeadows 

The pre-review build got a protocol mismatch going to dev domains, so the tester only tested in the sandbox (which succeeded). This was an error as the test plan calls for testing in sandbox and dev domains, not either/or. The tester should have mentioned the protocol issue before passing or failing the test.

conklin94122 commented:
@AndrewMeadows I'm going to set this back to QA Failed.

AndrewMeadows commented:
@conklin94122, QA_FAILED is fine but... what is this about protocol mismatch?  I am currently able to login to Dev-Welcome using the latest windows installer on this PR.

conklin94122 commented:
@AndrewMeadows The protocol mismatch was with an earlier build, not the latest. The build that was QA Passed got the mismatch.

AndrewMeadows commented:
I did a bit of testing to try to get some clues as to what is going wrong in Dev-Welcome (in this RP and in hifi/master).  Here are the symptoms I see (on both this PR and hifi/master):

Right after login when I run the test script which generates the test-sphere I see:

**(1)** The test-sphere is either non-colliding or it collides but is static.  Examination of entity properties shows the object as `dynamic=true`, `collisionless=false`, and that it should collide with avatars.  If I wait long enough the object will become collidable and dynamic.

**(2)** When the object is in state **(1)** if I click on it... nothing happens, but when the object becomes dynamic+collidable it will sometimes jump... but not always.  When it doesn't jump: `(a)` it is collisionless against the avatar, and yet its collision groups show it should collide, **(b)** when I click on it it moves (and collides with the floor, but remains collisionless with the avatar).  I believe this is caused by an orphaned "hold" or "grab" Action that is being put on the object when I click it too early: the test script did not set the non-grabbable properties soon enough and the orphaned Action is disabling collision with the avatar (a.k.a the anti-bootstrap feature).

**(3)** A minute or two after login, when I reload the script and create a new test-sphere things work as expected.

AndrewMeadows commented:
Most of the mysterious problems are somehow related to running **defaultScripts.js** and are not isolated to this PR: they also happen in `hifi-master`, so I've modified the test-plan: first disable the default scripts and only run the one from the test-plan.

We'll have to fix the other problems but not in this PR.

ZappoMan commented:
hmmm - disabling default scripts? that doesn't sound right.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6458/app_12185_PR.apk

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6458/HighFidelity-Beta-PR12185-75b5635d2fbf9c0fad8ee4a041061deeefc284d3.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6458/HighFidelity-Beta-PR12185-75b5635d2fbf9c0fad8ee4a041061deeefc284d3.dmg

QAadahlin commented:
Currently Testing

AndrewMeadows commented:
Ok after looking into this more I've learned some things:

(1) The "collisionless with avatar" problem was something else that was fixed in PR-12204.  I rebased this branch to pick up that fix.

(2) The test-sphere IS being added to the physics engine and can be dynamically moved in a timely manner, however its rendered appearance is NOT necessarily being updated in a timely manner at login in Dev-Welcome.  Also, it appears maybe that some signal/slot events are also slow to execute at login there.  Since the default scripts add to the work at login these problems are reduced by not running them.  This problem is not related to this PR.  I will probably provide a repro recipe that doesn't rely on what this PR is fixing and open a bug about it.

(3) The problem where sometimes the test-sphere rises and stops in mid-air is caused by something setting the object's gravity to zero.  When you examine the entity properties you'll see this to be the cause.  While I was able to reproduce this a few times I wasn't able to track down where it was coming from before it suddenly vanished.  It isn't clear to me if it was somehow fixed by one of the rebases I did while testing or if it just went into hiding.  However, I'm confident it isn't caused by this PR.

QAadahlin commented:
@conklin94122 Completed Testing for PR#12185. PASSED. Please note that in the domain Eschatology, the ball still gets stuck in mid-air. However, per @AndrewMeadows it is likely unrelated to the PR and has more to do with Eschatology domain's gravity.

AndrewMeadows commented:
Just an FYI, I was able to reproduce the floating ball problem in Eschatology using checkout of hifi/master prior to the merge of this PR.  Also, it happens even when not running defaultScripts.js.

The good news is that it looks like I have a semi-reliable repro recipe again.

AndrewMeadows commented:
Bah, I was only able to reproduce the zero-gravity bug using an older interface.  Using the tip of hifi/master it no longer happens.
